### DISCUSSION TITLE
how to call outer chain variable declaration(not from file directly)

### DISCUSSION THREAD

**Discussion by zrwsmd** [eclipse-langium/langium#1518] [Q&A]
i want to define a custom function block  `CacheFb1` at my st program var declaration,this custom function block  is not define at st file,i should get this from another way like a json file(this custom function block actually is a library function written by my colleague,he can give to me by json format)
![image](https://github.com/eclipse-langium/langium/assets/26995040/a5d59829-541c-4a75-b7b3-17a02461b01d)

for var declaration at st program ,it seems easy to achieve by changing my langium rule,but if i want to call it inner vaiable(`ccc` for example)  like below 
![image](https://github.com/eclipse-langium/langium/assets/26995040/74c6a325-fcad-4a9b-a589-3ec0dcb0dd0b)

i can not achieve it by override DefaultScopeProvider$getScope
so should i define $document property?maybe cache document?but it readonly property
![image](https://github.com/eclipse-langium/langium/assets/26995040/46997115-6e15-42c0-bd6c-38fc1f72856b)

**Comment by zrwsmd** [comment]
![image](https://github.com/eclipse-langium/langium/assets/26995040/242879a0-f212-4150-8460-c7372b0221f1)

**Comment by msujew** [comment]
Hey @zrwsmd,

This will work as expected once you correctly implement loading of external data. See https://github.com/eclipse-langium/langium/discussions/1517#discussioncomment-9591924.
